service/controller/azurecluster/handler/azureclusteridentity/create.go

```diff
@@ -0,0 +1,255 @@
+package azureclusteridentity
+
+import (
+	"context"
+	"reflect"
+
+	apiextensionslabels "github.com/giantswarm/apiextensions/v3/pkg/label"
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v5/pkg/project"
+	"github.com/giantswarm/azure-operator/v5/service/controller/key"
+)
+
+const (
+	secretDataFieldName = "clientSecret"
+
+	legacySecretClientIDFieldName     = "azure.azureoperator.clientid"
+	legacySecretClientSecretFieldName = "azure.azureoperator.clientsecret"
+	legacySecretTenantIDFieldName     = "azure.azureoperator.tenantid"
+)
+
+// EnsureCreated ensures there is an AzureClusterIdentity CR and a related Secret
+// with the same contents as the Giant Swarm credential secret.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	var err error
+	azureCluster, err := key.ToAzureCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Retrieve the legacy secret related to the organization this AzureCluster belongs to.
+	var legacySecret corev1.Secret
+	{
+		credentialSecret, err := r.azureClientsFactory.GetCredentialSecret(ctx, azureCluster.ObjectMeta)
```

Comment:
how much stuff would break if we'd change `GetCredentialSecret` to return a `Secret`? it feels weird to _get credential secret then find that secret_.

Reply:
should be doable I guess. Or we can create a new function in the interface just for this considering the whole authentication thing will change soonish when we'll work on https://github.com/giantswarm/giantswarm/issues/15602

Reply:
I took a look at this. I agree with you this would be a nice improvement, but since it would require a lot of changes I was thinking about leaving it as it is right now, and do a refactoring when we'll change the whole thing for ticket #15602.
Does it work for you or is this a blocker in your opinion?

Reply:
Works for me.